oeps/oep-0023-theming.rst

```diff
@@ -0,0 +1,204 @@
+      absolute minimum. If breaking changes must occur, the core
+      contributor team will enumerate them with every named release of
+      the Open edX platform.
+
+4. The core contributor team will provide official documentation,
+      examples, and tools to help other core contributors and open
+      source community members develop themes.
+
+5. Further modification to UI elements or content will not be officially
+      supported within the scope of this theming system. A system to
+      enable content modification may eventually be developed and
+      supported, but its definition is outside the scope of this
+      proposal.
+
+Plan
+----
+
+1. API definition for the new theming system will be proposed, agreed
+      upon, and documented.
+
+2. The core contributor team will develop an example theme compliant
+      with the new system and use it to build support into the platform.
+      Existing work [8] on such a theme already exists in the
+      edx-bootstrap repository.
+
+3. Open edX core and community contributors will build scaffolding to
+      support the new theming system throughout the Open edX platform.
+      This scaffolding will be implemented in as unobtrusive a means as
+      possible with all attempts made not to interfere with existing
+      theming systems.
+
+4. Once scaffolding is complete, the core contributor team will announce
+      an alpha release of the new theming system with the closest named
+      release of the Open edX platform.
+
+5. Throughout the lifecycle of the aforementioned release, the core
+      contributor team will solicit feedback and contributions and
+      respond to suggestions from community members willing to try out
+      the new theming system.
+
+6. By the next sequential named release of the Open edX platform,
+      assuming the new theming system has delivered value to its alpha
+      stakeholders, the core contributor team will announce a beta
+      release of the new theming system and announce the official
+      deprecation of all legacy theming systems.
+
+7. The Open edX core contributor team will continue to solicit
+      contributions and feedback on the theming system until the
+      following named release of the platform, by which point they will
+      announce official support for the new system.
+
+8. Removal of legacy theming-related code may begin two named releases
+      after the announcement of its deprecation.
+
+Consequences
+============
+
+-  All existing theming systems will be deprecated and support will
+      eventually be removed from the platform.
+
+-  Theming scope will be restricted to look-and-feel and branding
+      element modification. Content modification must take place outside
+      the scope of theming.
+
+-  The Open edX platform will gain further ability to evolve its
+      frontend architecture.
+
+-  The Open edX platform’s static asset size and build times will be
+      greatly reduced.
```

Comment:
Would this enable real-time theming (via https://github.com/medialize/sass.js) on near real-time (via background tasks)? It would be great if you could set some color values in the admin, click a button and immediately see the changes.

Reply:
Real-time theming comes up a lot, and while we agree it's ideal from an admin/maintainer experience standpoint, the technical complexity involved in enabling it for new *and* legacy systems is not a burden we currently have the resources to shoulder. CSS variables are promising, but would end support for IE11; CSS-in-JS is another (controversial!) option, but will not work with legacy non-JS UIs. A while back I even looked into building [an in-browser SASS compiler](https://github.com/arizzitano/theme-generator), but was unable to make much progress.

Like the current style theming system, this approach is SASS-based and will need to compile first; however, it involves a single scoped set of variables (typically in one file). We would certainly be open to implementing a real-time theming system in the future -- technical suggestions are welcome -- but I don't think it needs to be a blocker for approving Bootstrap adoption.

Reply:
@arizzitano Definitely not a blocker for Bootstrap adoption, but great to know that real-time theming has been under consideration! 

At OpenCraft we did [a short discovery on how a (near) real-time theming could work](https://docs.google.com/document/d/1WvjMZA2YNYKnM5acqT1rPOqCTo8JUDNsMumHL7baCis/edit?usp=sharing). It would be great if you could give it a quick look to confirm whether the proposed implementation makes sense.